File: tests/CoreTests/Delegates/Delegates.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SharpKit.JavaScript;

namespace CoreTests.Delegates
{
    [JsType(JsMode.Prototype, ForceDelegatesAsNativeFunctions=true)]
    class Delegates
    {
        void foo()
        {
            Action x = foo;
        }
    }


    public delegate string SomeDelegate();
    [JsType(JsMode.Prototype)]
    public class Class1
    {
        public Class1() { }

        public static SomeDelegate DelegateProp = delegate()
        {
            return "test";
        };
        public  SomeDelegate DelegateProp2 = delegate()
        {
            return "test";
        };
    }
}

File: tests/CoreTests/Delegates/Delegates.js
if (typeof ($CreateAnonymousDelegate) == 'undefined') {
    var $CreateAnonymousDelegate = function (target, func) {
        if (target == null || func == null)
            return func;
        var delegate = function () {
            return func.apply(target, arguments);
        };
        delegate.func = func;
        delegate.target = target;
        delegate.isDelegate = true;
        return delegate;
    }
}
if (typeof(CoreTests) == "undefined")
    var CoreTests = {};
if (typeof(CoreTests.Delegates) == "undefined")
    CoreTests.Delegates = {};
CoreTests.Delegates.Delegates = function ()
{
};
CoreTests.Delegates.Delegates.prototype.foo = function ()
{
    var x = this.foo;
};
CoreTests.Delegates.Class1 = function ()
{
    this.DelegateProp2 = $CreateAnonymousDelegate(this, function ()
    {
        return "test";
    });
};
CoreTests.Delegates.Class1.DelegateProp = function ()
{
    return "test";
};
